src/main/kotlin/org/rust/ide/injected/RsDoctestLanguageInjector.kt

```diff
@@ -0,0 +1,152 @@
+                }
+
+                // skip doc comment infix (`///`, `//!` or ` * `)
+                val ranges = lines.asSequence().mapNotNull { (lineStart, lineEnd) ->
+                    val index = text.indexOf(infix, lineStart)
+                    if (index != -1 && index < lineEnd) {
+                        val start = index + infix.length
+                        Triple(start, lineEnd, CharArrayUtil.shiftForward(text, start, lineEnd, " \t"))
+                    } else {
+                        null
+                    }
+                }.toList()
+
+                if (ranges.isEmpty()) return@mapNotNull null
+
+                val commonIndent = ranges.fold(Int.MAX_VALUE) { acc, (start, end, codeStart) ->
+                    Math.min(acc, if (codeStart + 1 == end) Int.MAX_VALUE else codeStart - start)
+                }
+
+                if (commonIndent > 0 && commonIndent != Int.MAX_VALUE) {
+                    ranges.mapNotNull { (start, end, codeStart) ->
+                        val newStart = start + commonIndent
+                        if (newStart < end) Triple(newStart, end, codeStart) else null
+                    }
+                } else {
+                    ranges
+                }.map { (start, end, codeStart) ->
+                    // `cargo doc` has special rules for code lines which start with `#`:
+                    //   * `# ` prefix is used to mark lines that should be skipped in rendered documentation.
+                    //   * `##` prefix is converted to `#`
+                    // See https://github.com/rust-lang/rust/blob/5182cc1c/src/librustdoc/html/markdown.rs#L114
+                    when {
+                        text.startsWith("##", codeStart) -> Pair(codeStart + 1, end)
+                        text.startsWith("# ", codeStart) -> Pair(codeStart + 2, end)
+                        else -> Pair(start, end)
+                    }
+                }
+            }.forEach { ranges ->
+                val inj = registrar.startInjecting(RsLanguage)
+                fun addPlace(start: Int, end: Int, first: Boolean, last: Boolean): MultiHostRegistrar {
+                    val prefix = if (first) {
+                        buildString {
+                            if (crateName != "std" || cargoTarget.pkg.origin != PackageOrigin.STDLIB) {
+                                append("extern crate ")
+                                append(crateName)
+                                append("; ")
+                            }
+                            append("fn main() {")
+                        }
+                    } else {
+                        null
+                    }
+                    val suffix = if (last) "}" else null
+                    return inj.addPlace(prefix, suffix, context, TextRange(start, end))
+                }
+                ranges.forEachIndexed { index, (start, end) ->
+                    val first = index == 0
```

Comment:
Also, it might be useful to introduce `CodeRange` instead of `Pair` and put the logic inside the class.
IMO it is difficult to read large `map`, `forEach`, and so forth.

Reply:
I've refactored this place a bit. Should be simpler now.